ensure remote server HTTP2 tests are actually using HTTP2

geoffkizer opened the pull request:
Fixes #39178 
Contributes to #34460

Update many remote server tests to take a RemoteServer argument (instead of just server uri) and use CreateHttpClientForRemoteServer to set the HttpClient's DefaultRequestVersion property based on the definition of the RemoteServer.

Also, validate during request processing that we are sending the correct request version and receiving the expected response version.

Also, update some remote server tests to run against the HTTP2 remote server as well.

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-linux

davidsh commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs:
```diff
@@ -190,23 +190,32 @@ public async Task UseCallback_NotSecureConnection_CallbackNotCalled()
+                        yield return new object[] {
+                            remoteServer,
+                            remoteServer.RedirectUriForDestinationUri(
+                                statusCode:302,
```
nit: add space after ":" for named parameters.

(comment applies to other places in the PR)

davidsh commented:
Windows 7 CI runs are failing with the HTTP/2.0 check. That's probably because Windows 7 doesn't support ALPN.

geoffkizer commented:
Yeah, makes sense. Will fix. Thanks.

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-linux

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-linux

geoffkizer commented:
@stephentoub Can you review?

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -224,11 +220,11 @@ public void Properties_AddItemToDictionary_ItemPresent()
         {
-            using (HttpClient client = CreateHttpClient())
-            using (HttpResponseMessage response = await client.GetAsync(remoteServer))
+            using (HttpClient client = CreateHttpClientForRemoteServer(remoteServer))
```
This HttpClientHandlerTest class has derivations for both 1.1:
https://github.com/dotnet/corefx/blob/290a6e54a482ee950c4b812ad85f8de2d795932f/src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs#L944
and 2.0:
https://github.com/dotnet/corefx/blob/290a6e54a482ee950c4b812ad85f8de2d795932f/src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs#L2232

In conjunction with this remote server approach, that then means we're running the same tests more than we should be? e.g. for both the 1.1 and 2.0 derivations, we'll run both 1.1 and 2.0 tests?

geoffkizer replied:
Yep. See #39345.

geoffkizer replied:
Note also that since a bunch of the tests in HttpClientHandlerTest are 1.1-loopback based (not generic loopback), we're also running these tests twice for 1.1, since in the "2.0" derivation they still run as 1.1.

The "2.0" derivation really means "use the 2.0 loopback server". It would be nice to clarify this, but it's not urgent.

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.cs:
```diff
@@ -39,7 +41,13 @@ public HttpClientHandlerTestBase(ITestOutputHelper output)
+            // Note that if you create the HttpRequestMessage explicitly, you will need to set its Version explicitly
+            // because it defaults to 1.1.
+
+            SetDefaultRequestVersion(client, HttpVersion.Version20);
```
I thought we were going to undo this?

geoffkizer replied:
Crap. Yes, I just forgot to. I'll remove it.

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.cs:
```diff
@@ -107,5 +115,58 @@ protected static object GetUnderlyingSocketsHttpHandler(HttpClientHandler handle
+
+                return response;
+            }
+        }
```
Neat.

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-linux

geoffkizer commented:
I'm seeing an assert in the certificate code on Linux. Seems to be happening for both CurlHandler and SocketsHttpHandler. @bartonjs any idea what's going on here?

```
Process terminated. Assertion failed.
Unrecognized X509VerifyStatusCode:44
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.MapVerifyErrorToChainStatus(X509VerifyStatusCode code) in /_/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:line 819
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.AddElementStatus(X509VerifyStatusCode errorCode, List`1 elementStatus, List`1 overallStatus) in /_/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:line 703
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.AddElementStatus(ErrorCollection errorCodes, List`1 elementStatus, List`1 overallStatus) in /_/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:line 694
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.BuildChainElements(WorkingChain workingChain, List`1& overallStatus) in /_/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:line 599
   at Internal.Cryptography.Pal.OpenSslX509ChainProcessor.Finish(OidCollection applicationPolicy, OidCollection certificatePolicy) in /_/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:line 393
   at Internal.Cryptography.Pal.ChainPal.BuildChain(Boolean useMachineContext, ICertificatePal cert, X509Certificate2Collection extraStore, OidCollection applicationPolicy, OidCollection certificatePolicy, X509RevocationMode revocationMode, X509RevocationFlag revocationFlag, DateTime verificationTime, TimeSpan timeout) in /_/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs:line 99
   at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate, Boolean throwOnException) in /_/src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Chain.cs:line 118
   at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate) in /_/src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Chain.cs:line 105
   at System.Net.Security.CertificateValidation.BuildChainAndVerifyProperties(X509Chain chain, X509Certificate2 remoteCertificate, Boolean checkCertName, String hostName) in /_/src/Common/src/System/Net/Security/CertificateValidation.Unix.cs:line 18
   at System.Net.CertificateValidationPal.VerifyCertificateProperties(SafeDeleteContext securityContext, X509Chain chain, X509Certificate2 remoteCertificate, Boolean checkCertName, Boolean isServer, String hostName) in /_/src/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs:line 22
   at System.Net.Security.SecureChannel.VerifyRemoteCertificate(RemoteCertValidationCallback remoteCertValidationCallback, ProtocolToken& alertToken) in /_/src/System.Net.Security/src/System/Net/Security/SecureChannel.cs:line 1010
   at System.Net.Security.SslStream.CompleteHandshake(ProtocolToken& alertToken) in /_/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs:line 754
```

Unfortuately, it's not obvious which test is causing this. Likely it's an existing test that didn't run over SSL before, but does now because of HTTP2.

bartonjs commented:
> any idea what's going on here?

It means OpenSSL has produced a code we've never seen before (and thus don't know how to map and don't know if it's fatal, reportable, or ignorable).  What distro (and version) are you on?

geoffkizer commented:
It looks like it's happening on every distro except RedHat.

https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F39343~2Fmerge/test~2Ffunctional~2Fcli~2Fouterloop~2F/20190710.2/workItem/System.Net.Http.Functional.Tests/wilogs

geoffkizer commented:
Previously, all HTTP2 tests were running with certificate validation disabled, because they HTTP2 derived test classes always disable this because the HTTP2 loopback server is using a fake cert. 

Given this, I'm going to disable the cert validation for HTTP2 on Linux and file a separate issue.

geoffkizer commented:
Filed #39372

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-linux

geoffkizer commented:
/azp run corefx-outerloop-osx

geoffkizer commented:
Interestingly, we're still hitting the same assert even with the code I added to try to skip cert validation on Linux for the HTTP2 endpoint.

I don't see anything wrong with the code I added, so this would seem to imply that we see this even with HTTP/1.1 over SSL. Not sure what to make of this. There must be a test that wasn't running over SSL before and now is, and it's hitting this assert. But all the tests are running against the same endpoint, so I have no idea why it would cause this assert now.

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-windows

geoffkizer commented:
/azp run corefx-outerloop-linux

geoffkizer commented:
/azp run corefx-outerloop-osx

geoffkizer commented:
/azp run corefx-outerloop-linux

geoffkizer commented:
/azp run

geoffkizer commented:
Failures are unrelated
